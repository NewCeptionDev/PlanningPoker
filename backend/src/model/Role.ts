export enum Role {
  PLAYER,
  OBSERVER,
  ADMIN,
}

export function roleFromString(role: string): Role {
  switch (role) {
    case 'PLAYER':
      return Role.PLAYER;
    case 'OBSERVER':
      return Role.OBSERVER;
    case 'ADMIN':
      return Role.ADMIN;
    default:
      return Role.PLAYER;
  }
}
