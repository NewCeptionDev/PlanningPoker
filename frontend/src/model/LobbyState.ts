export enum LobbyState {
  VOTING,
  OVERVIEW,
}
