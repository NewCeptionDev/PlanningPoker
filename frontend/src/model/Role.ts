export enum Role {
  PLAYER,
  OBSERVER,
  ADMIN,
}
