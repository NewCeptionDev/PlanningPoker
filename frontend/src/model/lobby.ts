export interface Lobby {
  users: User[];
  cardCollection: string[];
  state: LobbyState;
}
