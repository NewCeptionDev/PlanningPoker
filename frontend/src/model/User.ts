import { Role } from "./Role";

export interface User {
  id: string;
  name: string;
  cardSelected: boolean;
  selectedCard: string | undefined;
  roles: Role[];
}
