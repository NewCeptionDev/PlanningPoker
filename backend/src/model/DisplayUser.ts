import { Role } from './Role';
import { User } from './User';

export interface DisplayUser {
  id: string;
  name: string;
  cardSelected: boolean;
  selectedCard: string | undefined;
  roles: Role[];
}

export function toDisplayUser(
  user: User,
  fullInformation: boolean,
): DisplayUser {
  return {
    id: user.id,
    name: user.name,
    cardSelected: user.selectedCard !== undefined,
    selectedCard: fullInformation ? user.selectedCard : undefined,
    roles: user.roles,
  };
}
