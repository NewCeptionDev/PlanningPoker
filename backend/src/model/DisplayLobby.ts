import { DisplayUser } from './DisplayUser';
import { LobbyState } from './LobbyState';

export interface DisplayLobby {
  name: string;
  users: DisplayUser[];
  cardCollection: string[];
  state: LobbyState;
}
