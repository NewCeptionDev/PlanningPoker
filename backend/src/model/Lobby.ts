import { DisplayLobby } from './DisplayLobby'
import { LobbyState } from './LobbyState'
import { Option } from './Option'
import { Role } from './Role'
import { Socket } from 'socket.io'
import { User } from './User'

export class Lobby {
  id: string

  name: string

  users: User[]

  cardCollection: string[]

  state: LobbyState

  options: Option[]

  constructor(
    id: string,
    name: string,
    users: User[],
    cardCollection: string[],
    state: LobbyState,
    options: Option[],
  ) {
    this.id = id
    this.name = name
    this.users = users
    this.cardCollection = cardCollection
    this.state = state
    this.options = options

    // Add '?' as a card if it is not already included
    if (!this.cardCollection.includes('?')) {
      this.cardCollection.push('?')
    }
  }

  toDisplayLobby(fullInformation: boolean): DisplayLobby {
    return {
      name: this.name,
      users: this.users.map((u) => u.toDisplayUser(fullInformation)),
      cardCollection: this.cardCollection,
      state: this.state,
    }
  }

  toDisplayLobbyForUser(userId: string): DisplayLobby {
    return {
      name: this.name,
      users: this.users.map((u) =>
        u.toDisplayUser(u.id === userId || this.state === LobbyState.OVERVIEW),
      ),
      cardCollection: this.cardCollection,
      state: this.state,
    }
  }

  addUser(user: User) {
    if (this.users.find((u) => u.id === user.id)) {
      return
    }

    // The first User of a Lobby is always an Admin
    if (this.users.length === 0) {
      user.roles.push(Role.ADMIN)
    }

    this.users.push(user)
  }

  removeUser(socket: Socket): boolean {
    const user = this.users.find((u) => u.socketId === socket.id)
    if (!user) {
      return false
    }
    this.users = this.users.filter((u) => u.socketId !== socket.id)
    return true
  }

  resetSelectedCards() {
    this.users.forEach((u) => {
      u.selectCard(undefined)
    })
  }
}
