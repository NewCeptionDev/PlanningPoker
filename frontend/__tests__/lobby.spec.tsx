import LobbyScreen from '@/app/[slug]/lobby'
import { Role } from '@/model/Role'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

jest.mock('next/navigation', () => ({
  useRouter: () => {
    return {
      prefetch: () => {},
    }
  },
}))

describe('Lobby', () => {
  beforeAll(() => {
    // @ts-expect-error only partially defined
    const response: Response = {
      ok: true,
      json: async () => {
        return {
          lobbyExists: true,
          lobbyName: 'Test',
        }
      },
    }
    global.fetch = jest.fn().mockResolvedValue(response)
  })

  beforeEach(() => {
    jest.resetModules()
  })

  it('should render correctly', () => {
    const { container } = render(
      <LobbyScreen
        lobbyId='1234'
        user={{
          id: '1234',
          name: 'Test',
          roles: [Role.PLAYER],
          cardSelected: false,
          selectedCard: undefined,
        }}
      />,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render all important components', () => {
    render(
      <LobbyScreen
        lobbyId='1234'
        user={{
          id: '1234',
          name: 'Test',
          roles: [Role.PLAYER],
          cardSelected: false,
          selectedCard: undefined,
        }}
      />,
    )
    expect(screen.getByTestId('logo')).toBeInTheDocument()
    expect(screen.getByTestId('headline')).toBeInTheDocument()
    expect(screen.getByTestId('lobby-name')).toBeInTheDocument()
    expect(screen.getByTestId('theme-switcher')).toBeInTheDocument()
    expect(screen.getByTestId('lobby-info')).toBeInTheDocument()
    expect(screen.getByTestId('lobby-players')).toBeInTheDocument()
    expect(screen.getByTestId('lobby-users')).toBeInTheDocument()
    expect(screen.getByTestId('card-collection')).toBeInTheDocument()
  })
})
