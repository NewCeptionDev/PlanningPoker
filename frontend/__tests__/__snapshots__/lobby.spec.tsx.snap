// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lobby should render correctly 1`] = `
<div>
  <div
    class="flex flex-row h-[5vh] justify-between p-4 secondary"
  >
    <div
      class="w-1/5 flex items-center"
    >
      <div
        class="flex flex-row cursor-pointer"
      >
        <img
          alt="logo"
          class="mr-4"
          data-nimg="1"
          data-testid="logo"
          decoding="async"
          height="21"
          loading="lazy"
          src="/_next/image?url=%2Flogo.png&w=32&q=75"
          srcset="/_next/image?url=%2Flogo.png&w=16&q=75 1x, /_next/image?url=%2Flogo.png&w=32&q=75 2x"
          style="color: transparent;"
          width="16"
        />
        <h1
          class="text-white"
          data-testid="headline"
        >
          Planning Poker
        </h1>
      </div>
    </div>
    <div
      class="w-3/5 flex items-center justify-center"
    >
      <h1
        class="text-white"
        data-testid="lobby-name"
      />
    </div>
    <div
      class="flex flex-row justify-end w-1/5 items-center"
    >
      <h3
        class="text-white mr-8 ml-8"
      >
        Test
      </h3>
      <div
        class="grid place-items-center h-full relative"
        data-testid="theme-switcher"
      >
        <input
          class="hidden toggle--checkbox"
          data-testid="theme-switcher-toggle"
          id="toggle"
          type="checkbox"
        />
        <label
          class="toggle--label"
          data-testid="theme-switcher-label"
          for="toggle"
        >
          <span
            class="toggle--label-background"
            data-testid="theme-switcher-label-background"
          />
        </label>
      </div>
      <button
        class="btn ml-8"
      >
        Leave
      </button>
    </div>
  </div>
  <div
    class="flex flex-row h-[79vh] w-full"
  >
    <div
      class="w-1/5 flex flex-col m-8"
      data-testid="lobby-info"
    >
      <button
        class="btn w-3/4 mb-4"
      >
        Copy Invite Link
      </button>
      <h1>
        <b>
          Lobby:
        </b>
         
      </h1>
      <h3>
        <b>
          Id:
        </b>
         
        1234
      </h3>
    </div>
    <div
      class="w-3/5 flex flex-col items-center justify-evenly"
      data-testid="lobby-players"
    >
      <div
        class="h-1/4 w-1/2 flex flex-row justify-evenly items-end"
      />
      <div
        class="flex flex-row h-1/3 w-full justify-evenly"
      >
        <div
          class="h-full w-1/5 flex flex-col items-center justify-evenly"
        />
        <div
          class="border-4 w-1/2 h-full rounded-xl flex flex-col items-center justify-center border-secondary"
        >
          <p
            class="text-center"
          >
            Waiting for an admin to start the next round
          </p>
        </div>
        <div
          class="h-full w-1/5 flex flex-col items-center justify-evenly"
        />
      </div>
      <div
        class="h-1/4 w-1/2 flex flex-row justify-evenly items-end"
      />
    </div>
    <div
      class="w-1/5 flex flex-col items-center"
    >
      <div
        class="flex flex-col w-2/3"
        data-testid="lobby-users"
      >
        <h1
          class="text-center m-4"
        >
          <b>
            Connected Users
          </b>
        </h1>
      </div>
    </div>
  </div>
  <div
    class="flex flex-row justify-around items-center h-[16vh] ml-[5vw] mr-[5vw] overflow-hidden"
    data-testid="card-collection"
  />
</div>
`;
